File: Golang/Number of Operations to Make Network Connected.go
func makeConnected(n int, connections [][]int) int {
    if len(connections) < n-1 {
        return -1
    }
    g := make([][]int, n)
    for _, conn := range connections {
        g[conn[0]] = append(g[conn[0]], conn[1])
        g[conn[1]] = append(g[conn[1]], conn[0])
    }
    visited := make([]int, n)
    total := 0

    for i:=0; i < n; i++ {
        total = total + dfs(i, g, visited)
    }
    return total - 1
}

func dfs(node int, g [][]int, visited []int) int {
    if visited[node] == 1 {
        return 0
    }
    visited[node] = 1
    for _, i := range g[node] {
        dfs(i, g, visited)
    }
    return 1
}

File: Python3/Number of Operations to Make Network Connected.py
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:

        if len(connections) < n-1:
            return -1
        g = [set() for i in range(n)]
        for i, j in connections:
            g[i].add(j)
            g[j].add(i)
        visited = [0] * n

        def dfs(node):
            if visited[node]:
                return 0
            visited[node] = 1
            for i in g[node]:
                dfs(i)
            return 1
        
        return sum(dfs(i) for i in range(n)) - 1
            